### ISSUE TITLE
Padding arguments of ConvTranspose inconsistent for different backends

### ISSUE THREAD

**Issue by guillaumebaquiast** [keras-team/keras-core#774]
Running `examples/keras_io/vision/autoencoder.py` on a Torch backend currently fails, with the error `negative output_padding is not supported`. The problem doesn't happen for TF or JAX backends.

This is because the padding arguments of a Transposed Convolution have different conventions in different frameworks:
- TF uses `padding` (which is a string, "valid" or "same") to determine the initial shape of the output, and `output_padding` to increase the dimension of the output to the right.
- Torch uses `padding` (which is an int) to reduce the shape of the output both on the left and right sides, and `output_padding` to increase the dimension of the output to the right.

The current keras-core implementation tries to convert from one convention to the others (e.g., [here](https://github.com/keras-team/keras-core/blob/main/keras_core/backend/torch/nn.py#L467-L474)), but it seems like the conversion doesn't always work as expected.

After looking into, I actually *think* that it might not always be possible to convert from one convention to the other (see example at the end).
If that is correct, the only possibility that I see to ensure consistent results between different backends would be to not support the argument `output_padding` in ConvTranspose.

---
Example of incompatibility between TF and Torch conventions:

Doing a simple deconvolution of a vector `[1, 1, 1]` through a kernel `[1, 2]`, with `strides=2`. The "full" deconvolution output should be `[1, 2, 1, 2, 1, 2]`.

In TF, with `padding="same"` and `output_padding=1`, the output is `[1, 2, 1, 2, 1]` i.e., it "drops" the last element.
In Torch, with `padding=0` and `output_padding=0`, we have the "full" output: `[1, 2, 1, 2, 1, 2]`. Increasing `padding` would decrease the output *both* on the left and the right, while `output_padding` would increase just on the right. Therefore I don't see a way to get the same output as the one from TF.

If `output_padding` wasn't supported, this case wouldn't be possible: the result of the  from TF would be `[1, 2, 1, 2, 1, 2]`, just like the one from Torch.

**Comment by fchollet** [comment]
> After looking into, I actually think that it might not always be possible to convert from one convention to the other (see example at the end).

This may indeed be the case. There are existing discrepancies with average pooling ops as well due to padding.

> If that is correct, the only possibility that I see to ensure consistent results between different backends would be to not support the argument output_padding in ConvTranspose.

The argument is useful, so it would be better to restrict its usage while still enabling most use cases that need it. For instance, if you're trying to do something that isn't supported in torch, you should get an error message telling you exactly and offering alternatives.

**Comment by guillaumebaquiast** [comment]
> There are existing discrepancies with average pooling ops as well due to padding.

Interesting. Could you point to those discrepancies? I wonder if there is a general inconsistency on paddings, or if it's localised to a few operations.

> The argument is useful, so it would be better to restrict its usage while still enabling most use cases that need it. For instance, if you're trying to do something that isn't supported in torch, you should get an error message telling you exactly and offering alternatives.

Makes sense.
The padding inconsistencies probably don't matter much when training and inference are run within the same backend, as long as the output shapes are same (because during training the weights should adapt to the way the padding is defined). But they will matter when training using one backend and running inference with another one.
So, to confirm the definition of consistency between backends: a keras layer, initialised with the same arguments, and using the same weights, should return the same outputs (at inference time), irrespective of the backend. Correct?

**Comment by fchollet** [comment]
> Could you point to those discrepancies? I wonder if there is a general inconsistency on paddings, or if it's localised to a few operations.

Basically torch doesn't support asymmetric padding, so the last row/column in your average pooling outputs may have different values in torch (if the particular padding type and input shape led to asymmetric padding). It's similar to the new issue you're facing I suppose?

> The padding inconsistencies probably don't matter much when training and inference are run within the same backend, as long as the output shapes are same (because during training the weights should adapt to the way the padding is defined). But they will matter when training using one backend and running inference with another one.

Yes, same for the average pooling issue. This has led us to disable a keras applications architecture (I think MobileNet?) in PyTorch because the checkpoint was wrong for the backend.

Perhaps we could display a warning.

> So, to confirm the definition of consistency between backends: a keras layer, initialised with the same arguments, and using the same weights, should return the same outputs (at inference time), irrespective of the backend. Correct?

Ideally yes. So far the padding issue in torch is the only discrepancy we've had. This was listed in [the release FAQ](https://keras.io/keras_core/announcement/) (see question *"Do Keras Core models behave the same when run with different backends?"*)

**Comment by guillaumebaquiast** [comment]
Should be solved by https://github.com/keras-team/keras-core/pull/811
